### ISSUE TITLE
xiaoming90 - Price of Maverick LP tokens is vulnerable to manipulation

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-tokemak-judging#582]
xiaoming90

high

# Price of Maverick LP tokens is vulnerable to manipulation
## Summary

The price of the Maverick LP tokens is vulnerable to manipulation due to the use of spot reserves when computing the value.

## Vulnerability Detail

The `MavEthOracle` oracle obtains the spot reserves of the pool at Line 59 below and uses the following formula to compute the price of Maverick pool's LP token.

$$
price_{LP} = \frac{reserve_a \times price_a + reserve_b \times price_b}{L}
$$

where $reserve_i$ is token $i$ reserve amount, $price_i$ is the price of token $i$ in ETH, and $L$ is the total supply.

In short, it is calculated by getting all underlying balances, multiplying those by their market prices, and dividing them by the total supply of LP tokens. 

However, this approach of directly computing the price of LP tokens via spot reserves/balances is well-known to be vulnerable to manipulation.

https://github.com/sherlock-audit/2023-06-tokemak/blob/main/v2-core-audit-2023-07-14/src/oracles/providers/MavEthOracle.sol#L44

```solidity
File: MavEthOracle.sol
44:     function getPriceInEth(address _boostedPosition) external returns (uint256) {
..SNIP..
58:         // Get reserves in boosted position.
59:         (uint256 reserveTokenA, uint256 reserveTokenB) = boostedPosition.getReserves();
60: 
61:         // Get total supply of lp tokens from boosted position.
62:         uint256 boostedPositionTotalSupply = boostedPosition.totalSupply();
63: 
64:         IRootPriceOracle rootPriceOracle = systemRegistry.rootPriceOracle();
65: 
66:         // Price pool tokens.
67:         uint256 priceInEthTokenA = rootPriceOracle.getPriceInEth(address(pool.tokenA()));
68:         uint256 priceInEthTokenB = rootPriceOracle.getPriceInEth(address(pool.tokenB()));
69: 
70:         // Calculate total value of each token in boosted position.
71:         uint256 totalBoostedPositionValueTokenA = reserveTokenA * priceInEthTokenA;
72:         uint256 totalBoostedPositionValueTokenB = reserveTokenB * priceInEthTokenB;
73: 
74:         // Return price of lp token in boosted position.
75:         return (totalBoostedPositionValueTokenA + totalBoostedPositionValueTokenB) / boostedPositionTotalSupply;
76:         // slither-disable-end similar-names
77:     }
```

The reserves/balances of the pool can be manipulated, leading to fake LP token pricing. Following are additional references that highlight the issues with this formula.

- https://blog.alphaventuredao.io/fair-lp-token-pricing/
- https://docs.balancer.fi/reference/lp-tokens/valuing.html#directly-calculating-nav

The LP token is a pool position (PP) also called a boosted position, which is a distribution of liquidity within a given pool. LPs can join this PP by adding assets proportional to the existing PP asset mix. As such, anyone can add or remove liquidity to the PP, which will increase or decrease the underlying reserves to manipulate the price.

https://etherscan.io/address/0xa2b4e72a9d2d3252da335cb50e393f44a9f104ee#code#F23#L113

```solidity
function _getReserves(uint256 _tokenId) internal view checkBin returns (uint256 reserveA, uint256 reserveB) {
    uint256 binsLength = binIds.length;
    for (uint256 i; i < binsLength; i++) {
        (uint256 reserveA_, uint256 reserveB_, ) = _tokenBinReserves(_tokenId, i);
        reserveA += reserveA_;
        reserveB += reserveB_;
    }
}
```

## Impact

The protocol relies on the oracle to provide accurate pricing for many critical operations, such as determining the debt values of DV, calculators/stats used during the rebalancing process, NAV/shares of the LMPVault, and determining how much assets the users should receive during withdrawal. 

Incorrect pricing of LP tokens would result in many implications that lead to a loss of assets, such as users withdrawing more or fewer assets than expected due to over/undervalued vaults or strategy allowing an unprofitable rebalance to be executed.

## Code Snippet

https://github.com/sherlock-audit/2023-06-tokemak/blob/main/v2-core-audit-2023-07-14/src/oracles/providers/MavEthOracle.sol#L44

## Tool used

Manual Review

## Recommendation

Consider checking with Maverick's protocol team the best way to determine the price of a Pool Position's LP tokens in a manipulation-resistant manner.

Duplicate of #635
